File: test/lit/ts/trait.ts
// RUN: %cot emit-cir %s | %FileCheck %s

// Feature #057-058: Interface (protocol) declarations — TypeScript syntax
// TS `interface` with methods + `T extends Interface` generic constraint
// maps to cir.witness_table + cir.trait_call.

interface Summable {
    sum(): number;
}

function applySummable<T extends Summable>(val: T): number {
    return val.sum();
}

function main(): number {
    return 42;
}

// Generic function emitted with type_param and trait_call
// CHECK-LABEL: func.func @applySummable
// CHECK-SAME: !cir.type_param<"T">
// CHECK: cir.trait_call "Summable", "sum"
// CHECK: return

// Main function
// CHECK-LABEL: func.func @main
// CHECK: return

File: test/lit/swift/trait.swift
// RUN: %cot emit-cir %s | %FileCheck %s

// Feature #057-058: Protocol declarations + conformances — Swift syntax
// Swift protocol + extension maps to cir.witness_table + cir.trait_call.

struct Point {
    let x: Int32
    let y: Int32
}

protocol Summable {
    func sum() -> Int32
}

extension Point: Summable {
    func sum() -> Int32 {
        return self.x + self.y
    }
}

func applySummable<T: Summable>(_ val: T) -> Int32 {
    return val.sum()
}

func main() -> Int32 {
    return 42
}

// Impl method emitted with mangled name (self injected as first param)
// CHECK-LABEL: func.func @Point_Summable_sum
// CHECK-SAME: !cir.struct<"Point"
// CHECK: cir.field_val
// CHECK: cir.field_val
// CHECK: cir.add
// CHECK: return

// Witness table emitted
// CHECK: cir.witness_table "Point_Summable"
// CHECK-SAME: protocol("Summable")
// CHECK-SAME: methods(["sum"] = [@Point_Summable_sum])
